ICE `in operator(), at rust/typecheck/rust-hir-dot-operator.cc:285`

<!--
File: /tmp/im/3F8C937357066B8FA97BACDE01C5EB887B36C7D802ACAD4003B12617F7119E88.rs
-->

auto-reduced (treereduce-rust):
````rust
//@compile-flags: -frust-incomplete-and-experimental-compiler-do-not-use
trait A {
    fn a(&self) -> <Self as A>::X;
}

impl A for u32 {}

fn main() {
    let a: u32 = 0;
    let b: u32 = a.a();
}

````

original:
````rust
trait A {
    type Output;
    fn a(&self) -> <Self as A>::X;
    //~^ ERROR cannot find associated type `X` in trait `A`
}

impl A for u32 {
    type Output = u32;
    fn b(&self) -> u32 {
        0
    }
}

fn main() {
    let a: u32 = 0;
    let b: u32 = a.a();
}

````

Version information:

5dc8c4d20d62462ebce137c6e5146235ab9e9e2e

Possibly related line of code:
https://github.com/Rust-GCC/gccrs/blob/5dc8c4d20d62462ebce137c6e5146235ab9e9e2e/gcc/rust/typecheck/rust-hir-dot-operator.cc#L279-L291

Command:
`/home/matthias/vcs/github/gccrs/gccrs-build/gcc/crab1 -frust-incomplete-and-experimental-compiler-do-not-use`

<details><summary><strong>Program output</strong></summary>
<p>

```
/tmp/icemaker_global_tempdir.GdOY6g7Yxekn/rustc_testrunner_tmpdir_reporting.EPfSlJITfVyI/mvce.rs:2:33: error: cannot find associated type ‘X’ in trait ‘A’ [E0576]
    2 |     fn a(&self) -> <Self as A>::X;
      |                                 ^
      |                                 not found in `A`
/tmp/icemaker_global_tempdir.GdOY6g7Yxekn/rustc_testrunner_tmpdir_reporting.EPfSlJITfVyI/mvce.rs:2:5: error: failed to resolve return type
    2 |     fn a(&self) -> <Self as A>::X;
      |     ^~
/tmp/icemaker_global_tempdir.GdOY6g7Yxekn/rustc_testrunner_tmpdir_reporting.EPfSlJITfVyI/mvce.rs:5:1: error: missing a in implementation of trait ‘A’ [E0046]
    2 |     fn a(&self) -> <Self as A>::X;
      |     ~~
......
    5 | impl A for u32 {}
      | ^~~~
/tmp/icemaker_global_tempdir.GdOY6g7Yxekn/rustc_testrunner_tmpdir_reporting.EPfSlJITfVyI/mvce.rs:2:5: error: failed to resolve return type
    2 |     fn a(&self) -> <Self as A>::X;
      |     ^~
crab1: internal compiler error: in operator(), at rust/typecheck/rust-hir-dot-operator.cc:285
0x2fecaf8 internal_error(char const*, ...)
	../../gcc/diagnostic-global-context.cc:517
0xc0e04d fancy_abort(char const*, int, char const*)
	../../gcc/diagnostic.cc:1803
0x906d48 operator()
	../../gcc/rust/typecheck/rust-hir-dot-operator.cc:285
0x906d48 __invoke_impl<bool, Rust::Resolver::MethodResolver::assemble_trait_impl_candidates(const Rust::TyTy::BaseType&, std::vector<impl_item_candidate>&, std::vector<trait_item_candidate>&)::<lambda(Rust::HirId, Rust::HIR::ImplBlock*)>&, unsigned int, Rust::HIR::ImplBlock*>
	/usr/include/c++/15.1.1/bits/invoke.h:63
0x906d48 __invoke_r<bool, Rust::Resolver::MethodResolver::assemble_trait_impl_candidates(const Rust::TyTy::BaseType&, std::vector<impl_item_candidate>&, std::vector<trait_item_candidate>&)::<lambda(Rust::HirId, Rust::HIR::ImplBlock*)>&, unsigned int, Rust::HIR::ImplBlock*>
	/usr/include/c++/15.1.1/bits/invoke.h:116
0x906d48 _M_invoke
	/usr/include/c++/15.1.1/bits/std_function.h:292
0xf316ca std::function<bool (unsigned int, Rust::HIR::ImplBlock*)>::operator()(unsigned int, Rust::HIR::ImplBlock*) const
	/usr/include/c++/15.1.1/bits/std_function.h:593
0xf316ca Rust::Analysis::Mappings::iterate_impl_blocks(std::function<bool (unsigned int, Rust::HIR::ImplBlock*)>)
	../../gcc/rust/util/rust-hir-map.cc:831
0x10e897b Rust::Resolver::MethodResolver::assemble_trait_impl_candidates(Rust::TyTy::BaseType const&, std::vector<Rust::Resolver::MethodResolver::impl_item_candidate, std::allocator<Rust::Resolver::MethodResolver::impl_item_candidate> >&, std::vector<Rust::Resolver::MethodResolver::trait_item_candidate, std::allocator<Rust::Resolver::MethodResolver::trait_item_candidate> >&)
	../../gcc/rust/typecheck/rust-hir-dot-operator.cc:196
0x10eb818 Rust::Resolver::MethodResolver::select(Rust::TyTy::BaseType&)
	../../gcc/rust/typecheck/rust-hir-dot-operator.cc:427
0x1127211 Rust::Resolver::AutoderefCycle::try_autoderefed(Rust::TyTy::BaseType*)
	../../gcc/rust/typecheck/rust-autoderef.cc:422
0x11284ee Rust::Resolver::AutoderefCycle::cycle(Rust::TyTy::BaseType*)
	../../gcc/rust/typecheck/rust-autoderef.cc:328
0x10e8a68 Rust::Resolver::MethodResolver::Probe(Rust::TyTy::BaseType*, Rust::HIR::PathIdentSegment const&, bool)
	../../gcc/rust/typecheck/rust-hir-dot-operator.cc:40
0x10a1632 Rust::Resolver::TypeCheckExpr::visit(Rust::HIR::MethodCallExpr&)
	../../gcc/rust/typecheck/rust-hir-type-check-expr.cc:1307
0x109a4d4 Rust::Resolver::TypeCheckExpr::Resolve(Rust::HIR::Expr&)
	../../gcc/rust/typecheck/rust-hir-type-check-expr.cc:50
0x10aa071 Rust::Resolver::TypeCheckStmt::visit(Rust::HIR::LetStmt&)
	../../gcc/rust/typecheck/rust-hir-type-check-stmt.cc:90
0x10aa163 Rust::Resolver::TypeCheckStmt::Resolve(Rust::HIR::Stmt&)
	../../gcc/rust/typecheck/rust-hir-type-check-stmt.cc:34
0x109bef1 Rust::Resolver::TypeCheckExpr::visit(Rust::HIR::BlockExpr&)
	../../gcc/rust/typecheck/rust-hir-type-check-expr.cc:609
0x109a4d4 Rust::Resolver::TypeCheckExpr::Resolve(Rust::HIR::Expr&)
	../../gcc/rust/typecheck/rust-hir-type-check-expr.cc:50
0x10773f0 Rust::Resolver::TypeCheckItem::visit(Rust::HIR::Function&)
	../../gcc/rust/typecheck/rust-hir-type-check-item.cc:642
Please submit a full bug report, with preprocessed source (by using -freport-bug).
Please include the complete backtrace with any bug report.
See <https://gcc.gnu.org/bugs/> for instructions.

```

</p>
</details>

--- a/gcc/rust/typecheck/rust-hir-dot-operator.cc
+++ b/gcc/rust/typecheck/rust-hir-dot-operator.cc
@@ -51,6 +51,9 @@ MethodResolver::Select (std::set<MethodCandidate> &candidates,
     {
       TyTy::BaseType *candidate_type = candidate.candidate.ty;
       rust_assert (candidate_type->get_kind () == TyTy::TypeKind::FNDEF);
+      if (candidate_type == nullptr
+	  || candidate_type->get_kind () != TyTy::TypeKind::FNDEF)
+	continue;
       TyTy::FnType &fn = *static_cast<TyTy::FnType *> (candidate_type);
 
       // match the number of arguments
@@ -136,11 +139,11 @@ MethodResolver::assemble_inherent_impl_candidates (
       TyTy::BaseType *ty = nullptr;
       if (!query_type (func->get_mappings ().get_hirid (), &ty))
 	return true;
-      rust_assert (ty != nullptr);
-      if (ty->get_kind () == TyTy::TypeKind::ERROR)
+      if (ty == nullptr || ty->get_kind () == TyTy::TypeKind::ERROR)
+	return true;
+      if (ty->get_kind () != TyTy::TypeKind::FNDEF)
 	return true;
 
-      rust_assert (ty->get_kind () == TyTy::TypeKind::FNDEF);
       TyTy::FnType *fnty = static_cast<TyTy::FnType *> (ty);
       const TyTy::BaseType *impl_self
 	= TypeCheckItem::ResolveImplBlockSelf (*impl);
@@ -220,10 +223,11 @@ MethodResolver::assemble_trait_impl_candidates (
 	TyTy::BaseType *ty = nullptr;
 	if (!query_type (func->get_mappings ().get_hirid (), &ty))
 	  continue;
-	if (ty->get_kind () == TyTy::TypeKind::ERROR)
+	if (ty == nullptr || ty->get_kind () == TyTy::TypeKind::ERROR)
+	  continue;
+	if (ty->get_kind () != TyTy::TypeKind::FNDEF)
 	  continue;
 
-	rust_assert (ty->get_kind () == TyTy::TypeKind::FNDEF);
 	TyTy::FnType *fnty = static_cast<TyTy::FnType *> (ty);
 	const TyTy::BaseType *impl_self
 	  = TypeCheckItem::ResolveImplBlockSelf (*impl);
@@ -282,7 +286,8 @@ MethodResolver::assemble_trait_impl_candidates (
       return true;
 
     TyTy::BaseType *ty = item_ref->get_tyty ();
-    rust_assert (ty->get_kind () == TyTy::TypeKind::FNDEF);
+    if (ty == nullptr || ty->get_kind () != TyTy::TypeKind::FNDEF)
+      return true;
     TyTy::FnType *fnty = static_cast<TyTy::FnType *> (ty);
 
     trait_candidates.emplace_back (func, trait, fnty, trait_ref, item_ref);
@@ -298,7 +303,8 @@ MethodResolver::try_select_predicate_candidates (TyTy::BaseType &receiver)
   for (const auto &predicate : predicate_items)
     {
       const TyTy::FnType *fn = predicate.fntype;
-      rust_assert (fn->is_method ());
+      if (!fn->is_method ())
+	continue;
 
       TyTy::BaseType *fn_self = fn->get_self_type ();
       rust_debug ("dot-operator predicate fn_self={%s} can_eq receiver={%s}",
@@ -345,7 +351,8 @@ MethodResolver::try_select_inherent_impl_candidates (
 	continue;
 
       TyTy::FnType *fn = impl_item.ty;
-      rust_assert (fn->is_method ());
+      if (!fn->is_method ())
+	continue;
 
       TyTy::BaseType *fn_self = fn->get_self_type ();
 
@@ -383,7 +390,8 @@ MethodResolver::try_select_trait_impl_candidates (
   for (auto trait_item : candidates)
     {
       TyTy::FnType *fn = trait_item.ty;
-      rust_assert (fn->is_method ());
+      if (!fn->is_method ())
+	continue;
 
       TyTy::BaseType *fn_self = fn->get_self_type ();
       rust_debug ("dot-operator trait_item fn_self={%s} can_eq receiver={%s}",

--- a/gcc/testsuite/rust/compile/issue-3958.rs
+++ b/gcc/testsuite/rust/compile/issue-3958.rs
@@ -1,0 +1,11 @@
+// { dg-options "-fsyntax-only" }
+trait A {
+    fn a(&self) -> <Self as A>::X;
+}
+
+impl A for u32 {}
+
+fn main() {
+    let a: u32 = 0;
+    let b: u32 = a.a();
+}

